virtual_ecosystem/core/base_model.py

```diff
@@ -240,24 +241,142 @@ def __init__(
+
+        Raises:
+            ConfigurationError: If the model is static and some but not all the
+            variables or None in vars_populated_by_init are present in the data object
+            or if the model is not static and some of the variables in
+            vars_populated_by_init are not present in the data object.
+
+        Returns:
+            True if the model is static and all variables are present, such that the
+            setup method can be bypassed. False otherwise.
+        """
+        present = [var for var in self.vars_populated_by_init if var in self.data]
+        found = len(present)
+        expected = len(self.vars_populated_by_init)
+
+        if not self._static and present:
+            raise ConfigurationError(
+                f"Non-static model {self.model_name} requires none of the variables in "
+                f"vars_populated_by_init to be present in the data object. "
+                f"Present variables: {present}"
+            )
+
+        elif self._static:
+            if 0 < found < expected:
+                raise ConfigurationError(
+                    f"Static model {self.model_name} requires to either all variables "
+                    "in vars_populated_by_init to be present in the data object or "
+                    f"all to be absent. {found} out of {expected} found: {present}."
+                )
+            else:
+                return True
+
+        return False
+
+    def _run_update_due_to_static_configuration(self) -> bool:
+        """Decides if the update should be bypassed based on the static flag.
+
+        In particular, it checks that the appropriate variables created or updated in
+        the update method are present or not in the data object. Based on this, an
+        exception is raised is there is a problem or a decission is made on whether or
+        not running the update method once.
+
+        Raises:
+            ConfigurationError: If the model is static and some but not all the
+            variables or None in vars_populated_by_first_update or vars_updated are
+            present in the data object or if the model is not static and some of the
+            variables in vars_populated_by_init are not present in the data object.
+
+        Returns:
+            True if the model is static and the update method needs to run once. False
+            otherwise.
+        """
+        required = set(self.vars_populated_by_first_update + self.vars_updated)
+        present = [var for var in required if var in self.data]
+        found = len(present)
+        expected = len(required)
+
+        if not self._static and present:
+            raise ConfigurationError(
+                f"Non-static model {self.model_name} requires none of the variables in "
+                f"vars_populated_by_first_update or vars_updated to be present in the "
+                f"data object. Present variables: {present}"
```

Comment:
The message here seems different from the way the error is described in `Raises:` section of the docstring (I might be missing something). This message suggests the error is because none of the variables in "vars_populated_by_first_update or vars_updated" are required to be in the data object whereas the docstring says that non-static models should only error if "some of the variables in vars_populated_by_init are not present in the data object"

Reply:
Yep, the docstring is wrong. The result of copy/paste. Sorry.

Reply:
Ahh that makes more sense! So is it the case that my models (litter and soil) should add things to `vars_updated` that are also included in `vars_required_for_init`?

Reply:
No. In `vars_updated` you only need to include things that are updated by the model `update` method. If the model is not static (which is the default), then these variables should NOT be in the data object when setting up the model - i.e. they should not be provided as input to a simulation - which is what this error is complaining about.

There are two separate checks, on in relation to the `vars_required_for_init` a few lines above and another one related to the `vars_populated_by_first_update` and `vars_updated` (this check). Both are run when running `__init__`.

Reply:
Ahh right! So for example, if `"soil_c_pool_microbe"` is needed for init (to set the initial condition) and then gets continually updated by update method. Would it need to be included in both `vars_required_for_init` and `vars_updated` (which is what I'm currently doing), or just be included in `vars_required_for_init`?

Reply:
In that case, it should appear in both groups, as it fits both conditions: it is required for init and, as well, it is updated every iteration.

Reply:
Now, this flags a problem: you need the variable to be in the data object, so the model can be initialised, but at the same time you do not want it here otherwise it will cause an error - as it is happening. 

I guess the solution might be to include in this check only the variables initialised by the first update and not those updated by init, since these might be required to initialise the model. Or exclude the variables that are both in required by init and vars_updated.

Reply:
I guess either approach would work to fix the problem (don't have a strong opinion on which is preferable)